auth/saml/saml.go

```diff
@@ -0,0 +1,381 @@
+	PublicCert            string
+	PrivateKey            string
+	IdpUrl                string
+	IdpPublicCert         string
+	SignRequest           bool
+	SignedResponse        bool
+	DeflatEncodedResponse bool
+}
+
+func init() {
+	auth.RegisterScheme("saml", &SAMLAuthScheme{})
+}
+
+func (s SAMLAuthScheme) AppLogout(token string) error {
+	return s.Logout(token)
+}
+
+// This method loads basic config and returns a copy of the
+// config object.
+func (s *SAMLAuthScheme) loadConfig() (BaseConfig, error) {
+	if s.BaseConfig.EntityID != "" {
+		return s.BaseConfig, nil
+	}
+	if s.Parser == nil {
+		s.Parser = s
+	}
+	var emptyConfig BaseConfig
+	publicCert, err := config.GetString("auth:saml:sp-publiccert")
+	if err != nil {
+		return emptyConfig, err
+	}
+	privateKey, err := config.GetString("auth:saml:sp-privatekey")
+	if err != nil {
+		return emptyConfig, err
+	}
+	idpUrl, err := config.GetString("auth:saml:idp-ssourl")
+	if err != nil {
+		return emptyConfig, err
+	}
+	displayName, err := config.GetString("auth:saml:sp-display-name")
+	if err != nil {
+		displayName = "Tsuru"
+		log.Debugf("auth:saml:sp-display-name not found using default: %s", err)
+	}
+	description, err := config.GetString("auth:saml:sp-description")
+	if err != nil {
+		description = "Tsuru Platform as a Service software"
+		log.Debugf("auth:saml:sp-description not found using default: %s", err)
+	}
+	idpPublicCert, err := config.GetString("auth:saml:idp-publiccert")
+	if err != nil {
+		return emptyConfig, err
+	}
+	entityId, err := config.GetString("auth:saml:sp-entityid")
+	if err != nil {
+		return emptyConfig, err
+	}
+	signRequest, err := config.GetBool("auth:saml:sp-sign-request")
+	if err != nil {
+		return emptyConfig, err
+	}
+	signedResponse, err := config.GetBool("auth:saml:idp-sign-response")
+	if err != nil {
+		return emptyConfig, err
+	}
+	deflatEncodedResponse, err := config.GetBool("auth:saml:idp-deflate-encoding")
+	if err != nil {
+		deflatEncodedResponse = false
+		log.Debugf("auth:saml:idp-deflate-encoding not found using default [false]: %s", err)
+	}
+	s.BaseConfig = BaseConfig{
+		EntityID:              entityId,
+		DisplayName:           displayName,
+		Description:           description,
+		PublicCert:            publicCert,
+		PrivateKey:            privateKey,
+		IdpUrl:                idpUrl,
+		IdpPublicCert:         idpPublicCert,
+		SignRequest:           signRequest,
+		SignedResponse:        signedResponse,
+		DeflatEncodedResponse: deflatEncodedResponse,
+	}
+	return s.BaseConfig, nil
+}
+
+func Metadata() (string, error) {
+	scheme := SAMLAuthScheme{}
+	sp, err := scheme.createSP()
+	if err != nil {
+		return "", err
+	}
+	md, err := sp.GetEntityDescriptor()
+	if err != nil {
+		return "", err
+	}
+	return md, nil
+}
+
+func (s *SAMLAuthScheme) Login(params map[string]string) (auth.Token, error) {
+	_, err := s.loadConfig()
+	if err != nil {
+		return nil, err
+	}
+	//verify for callback requests, param 'callback' indicate callback
```

Comment:
Can you remove this comment?

Also, I think we can make the whole condition smaller by combining if and assignment:

```golang
if _, ok := params["callback"]; ok {
    return nil, s.callback(params)
}
```

What do you think?